proposals/003-authentication-api.md

```diff
@@ -0,0 +1,647 @@
+     * This method may be called multiple times over the lifetime of
+     * a session if reauthentication is required.
+     * TODO define the semantics around reauth
+     * @param saslPrincipal The authenticated principal.
+     */
+    void serverSaslAuthenticationSuccess(String mechanism, String serverName);
+
+    /**
+     * Allows a filter
+     * to announce a failed authentication outcome with the Kafka server.
+     * @param exception An exception describing the authentication failure.
+     */
+    void serverSaslAuthenticationFailure(String mechanism, String serverName, Exception exception);
+```
+
+### APIs for Filters to consume server SASL information
+
+```java
+    /**
+     * @return The SASL context for the server connection, or empty if the server
+     * has not successfully authenticated using SASL.
+     */
+    Optional<ServerSaslContext> serverSaslContext();
+```
+
+Where:
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+import java.util.Optional;
+
+import io.kroxylicious.proxy.filter.FilterContext;
+
+/**
+ * Exposes SASL authentication information to plugins, for example using {@link FilterContext#serverSaslContext()} ()}.
+ * This is implemented by the runtime for use by plugins.
+ */
+public interface ServerSaslContext {
+
+    /**
+     * The name of the SASL mechanism used.
+     * @return The name of the SASL mechanism used.
+     */
+    String mechanismName();
+}
+```
+
+### APIs for using server principals in a generic way
+
+This works similarly to the client-facing equivalent:
+
+```java
+    /**
+     * Returns the authenticated principal for the server connection, or empty if the server
+     * has not successfully authenticated, or if the server authentication was not mutual.
+     * The concrete type of principal returned depends on the proxy configuration.
+     * For example,
+     * it may be a {@link javax.security.auth.x500.X500Principal} if server identity is TLS-based,
+     * or it may be a {@link SaslPrincipal} is client identity is SASL based.
+     * @return The authenticated principal for the server connection, or empty if the server
+     * has not successfully authenticated.
+     */
+    Optional<? extends Principal> serverPrincipal();
+```
+
+
+### API for selecting target cluster TLS credentials
+
+The APIs presented so far are sufficient to write plugins which:
+
+* Propagate SASL, letting a client's SASL identity reach the server unchanged
+* Initiate SASL, injecting mechanism-specific credentials prior to letting client-originated requests reach the server. 
+  The selection of the server-facing SASL credentials to use could be based on the client's identity 
+  (e.g. SASL termination and SASL initiation in the same virtual cluster)
+
+What's missing is an API where the server-facing TLS client certificate is chosen based on the client's identity. 
+For this purpose we will add a new plugin interface, `ServerTlsCredentialSupplierFactory`.
```

Comment:
@robobario I was just looking at 0.14.0 and did not find a way/plugin to propagate TLS credentials to upstream. Having the ClientTlsContext in the Filter context is great! We can add our custom filtering logic that way however we would not be able to then propagate same creds upstream so that the brokers can use those creds for ACLs enforcement, throttling etc.

Is this api still being considered?

Reply:
@kidpollo that part of this proposal got factored into https://github.com/kroxylicious/design/pull/75. I'd still like to implement that, but I've been working on other things recently, so I don't have an implementation. I'd like there to be at least a partial implementation before we commit to that proposal.